## Extraccion de Datos

Librerias

In [27]:
import re
import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt

In [8]:
# Abrimos el DataSet entregado por Henrry Homicidios hoja Hechos

homicidios_hechos = pd.read_excel(r"C:\Users\ayrc2\Documentos\Proyecto Individual DPT03\PI_02_DA_SINIESTROS_VIALES\PI_02_DA_Siniestros_Viales\Data\homicidios.xlsx", sheet_name="HECHOS")
homicidios_hechos.head()

ID  N_VICTIMAS      FECHA  AAAA  MM  DD      HORA HH  \
0  2016-0001           1 2016-01-01  2016   1   1  04:00:00  4   
1  2016-0002           1 2016-01-02  2016   1   2  01:15:00  1   
2  2016-0003           1 2016-01-03  2016   1   3  07:00:00  7   
3  2016-0004           1 2016-01-10  2016   1  10  00:00:00  0   
4  2016-0005           1 2016-01-21  2016   1  21  05:20:00  5   

                             LUGAR_DEL_HECHO TIPO_DE_CALLE  ...  Altura  \
0  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA  ...     NaN   
1          AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ  ...     NaN   
2                         AV ENTRE RIOS 2034       AVENIDA  ...  2034.0   
3      AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA  ...     NaN   
4  AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA  ...     NaN   

                                 Cruce  \
0  FERNANDEZ DE LA CRUZ, F., GRAL. AV.   
1                  DE LOS CORRALES AV.   
2                                  NaN   
3             VILLEGAS, CONRADO, GRAL.   
4              SAENZ PE?A, LUIS, PRES.   

                               Dirección Normalizada COMUNA  \
0  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...      8   
1               PAZ, GRAL. AV. y DE LOS CORRALES AV.      9   
2                                ENTRE RIOS AV. 2034      1   
3          LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.      8   
4            SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.      1   

                                 XY (CABA)         pos x         pos y  \
0    Point (98896.78238426 93532.43437792)  -58.47533969  -34.68757022   
1    Point (95832.05571093 95505.41641999)  -58.50877521  -34.66977709   
2   Point (106684.29090040 99706.57687843)  -58.39040293  -34.63189362   
3    Point (99840.65224780 94269.16534422)  -58.46503904  -34.68092974   
4  Point (106980.32827929 100752.16915795)  -58.38718297  -34.62246630   

    PARTICIPANTES VICTIMA    ACUSADO  
0       MOTO-AUTO    MOTO       AUTO  
1  AUTO-PASAJEROS    AUTO  PASAJEROS  
2       MOTO-AUTO    MOTO       AUTO  
3         MOTO-SD    MOTO         SD  
4  MOTO-PASAJEROS    MOTO  PASAJEROS  

[5 rows x 21 columns]

In [11]:
#validamos que columnas tiene
homicidios_hechos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 696 entries, 0 to 695
Data columns (total 21 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   ID                     696 non-null    object        
 1   N_VICTIMAS             696 non-null    int64         
 2   FECHA                  696 non-null    datetime64[ns]
 3   AAAA                   696 non-null    int64         
 4   MM                     696 non-null    int64         
 5   DD                     696 non-null    int64         
 6   HORA                   696 non-null    object        
 7   HH                     696 non-null    object        
 8   LUGAR_DEL_HECHO        696 non-null    object        
 9   TIPO_DE_CALLE          696 non-null    object        
 10  Calle                  695 non-null    object        
 11  Altura                 129 non-null    float64       
 12  Cruce                  525 non-null    object        
 13  Direc

In [12]:
# Abrimos el DataSet entregado por Henrry Homicidios hoja Victimas

homicidios_victimas = pd.read_excel(r"C:\Users\ayrc2\Documentos\Proyecto Individual DPT03\PI_02_DA_SINIESTROS_VIALES\PI_02_DA_Siniestros_Viales\Data\homicidios.xlsx", sheet_name="VICTIMAS")
homicidios_victimas.head()

ID_hecho      FECHA  AAAA  MM  DD        ROL VICTIMA       SEXO EDAD  \
0  2016-0001 2016-01-01  2016   1   1  CONDUCTOR    MOTO  MASCULINO   19   
1  2016-0002 2016-01-02  2016   1   2  CONDUCTOR    AUTO  MASCULINO   70   
2  2016-0003 2016-01-03  2016   1   3  CONDUCTOR    MOTO  MASCULINO   30   
3  2016-0004 2016-01-10  2016   1  10  CONDUCTOR    MOTO  MASCULINO   18   
4  2016-0005 2016-01-21  2016   1  21  CONDUCTOR    MOTO  MASCULINO   29   

   FECHA_FALLECIMIENTO  
0  2016-01-01 00:00:00  
1  2016-01-02 00:00:00  
2  2016-01-03 00:00:00  
3                   SD  
4  2016-02-01 00:00:00

In [13]:
#validamos que columnas tiene
homicidios_victimas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 717 entries, 0 to 716
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   ID_hecho             717 non-null    object        
 1   FECHA                717 non-null    datetime64[ns]
 2   AAAA                 717 non-null    int64         
 3   MM                   717 non-null    int64         
 4   DD                   717 non-null    int64         
 5   ROL                  717 non-null    object        
 6   VICTIMA              717 non-null    object        
 7   SEXO                 717 non-null    object        
 8   EDAD                 717 non-null    object        
 9   FECHA_FALLECIMIENTO  717 non-null    object        
dtypes: datetime64[ns](1), int64(3), object(6)
memory usage: 56.1+ KB


## Transformacion de la data `homicidios_hechos`

In [14]:
homicidios_hechos

ID  N_VICTIMAS      FECHA  AAAA  MM  DD      HORA  HH  \
0    2016-0001           1 2016-01-01  2016   1   1  04:00:00   4   
1    2016-0002           1 2016-01-02  2016   1   2  01:15:00   1   
2    2016-0003           1 2016-01-03  2016   1   3  07:00:00   7   
3    2016-0004           1 2016-01-10  2016   1  10  00:00:00   0   
4    2016-0005           1 2016-01-21  2016   1  21  05:20:00   5   
..         ...         ...        ...   ...  ..  ..       ...  ..   
691  2021-0093           1 2021-12-13  2021  12  13  17:10:00  17   
692  2021-0094           1 2021-12-20  2021  12  20  01:10:00   1   
693  2021-0095           1 2021-12-30  2021  12  30  00:43:00   0   
694  2021-0096           1 2021-12-15  2021  12  15  10:30:00  10   
695  2021-0097           1 2021-11-18  2021  11  18  06:10:00   6   

                               LUGAR_DEL_HECHO TIPO_DE_CALLE  ...  Altura  \
0    AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA  ...     NaN   
1            AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ  ...     NaN   
2                           AV ENTRE RIOS 2034       AVENIDA  ...  2034.0   
3        AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA  ...     NaN   
4    AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA  ...     NaN   
..                                         ...           ...  ...     ...   
691                          AV. RIESTRA Y MOM       AVENIDA  ...     NaN   
692                    AU DELLEPIANE Y LACARRA     AUTOPISTA  ...     NaN   
693                        AV. GAONA Y TERRADA       AVENIDA  ...     NaN   
694                         AV. EVA PERON 4071       AVENIDA  ...  4071.0   
695                    PADRE CARLOS MUJICA 709         CALLE  ...   709.0   

                                   Cruce  \
0    FERNANDEZ DE LA CRUZ, F., GRAL. AV.   
1                    DE LOS CORRALES AV.   
2                                    NaN   
3               VILLEGAS, CONRADO, GRAL.   
4                SAENZ PE?A, LUIS, PRES.   
..                                   ...   
691                                  MOM   
692                          LACARRA AV.   
693                              TERRADA   
694                                  NaN   
695                                  NaN   

                                 Dirección Normalizada COMUNA  \
0    PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...      8   
1                 PAZ, GRAL. AV. y DE LOS CORRALES AV.      9   
2                                  ENTRE RIOS AV. 2034      1   
3            LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.      8   
4              SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.      1   
..                                                 ...    ...   
691                                  RIESTRA AV. y MOM      7   
692         DELLEPIANE, LUIS, TTE. GRAL. y LACARRA AV.      9   
693                                GAONA AV. y TERRADA     11   
694                                PERON, EVA AV. 4071      9   
695                            PADRE CARLOS MUGICA 709      1   

                                   XY (CABA)         pos x         pos y  \
0      Point (98896.78238426 93532.43437792)  -58.47533969  -34.68757022   
1      Point (95832.05571093 95505.41641999)  -58.50877521  -34.66977709   
2     Point (106684.29090040 99706.57687843)  -58.39040293  -34.63189362   
3      Point (99840.65224780 94269.16534422)  -58.46503904  -34.68092974   
4    Point (106980.32827929 100752.16915795)  -58.38718297  -34.62246630   
..                                       ...           ...           ...   
691   Point (102728.60090138 98186.24929177)  -58.43353773  -34.64561636   
692    Point (99624.29795829 97569.69801131)  -58.46739825  -34.65117757   
693   Point (99116.45492358 101045.23284826)  -58.47293407  -34.61984745   
694    Point (99324.54463985 97676.26932409)  -58.47066794  -34.65021673   
695  Point (107664.16647795 104708.63962087)  -58.37976155  -34.58679619   

      PARTICIPANTES    VICTIMA    A

In [15]:
#validamos que columnas tiene
homicidios_hechos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 696 entries, 0 to 695
Data columns (total 21 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   ID                     696 non-null    object        
 1   N_VICTIMAS             696 non-null    int64         
 2   FECHA                  696 non-null    datetime64[ns]
 3   AAAA                   696 non-null    int64         
 4   MM                     696 non-null    int64         
 5   DD                     696 non-null    int64         
 6   HORA                   696 non-null    object        
 7   HH                     696 non-null    object        
 8   LUGAR_DEL_HECHO        696 non-null    object        
 9   TIPO_DE_CALLE          696 non-null    object        
 10  Calle                  695 non-null    object        
 11  Altura                 129 non-null    float64       
 12  Cruce                  525 non-null    object        
 13  Direc

Nulos y Duplicados

In [16]:
#Revisamos las columnas con datos Nulos
homicidios_hechos.isnull().sum()

ID                         0
N_VICTIMAS                 0
FECHA                      0
AAAA                       0
MM                         0
DD                         0
HORA                       0
HH                         0
LUGAR_DEL_HECHO            0
TIPO_DE_CALLE              0
Calle                      1
Altura                   567
Cruce                    171
Dirección Normalizada      8
COMUNA                     0
XY (CABA)                  0
pos x                      0
pos y                      0
PARTICIPANTES              0
VICTIMA                    0
ACUSADO                    0
dtype: int64

* Para la columna Calle tiene 1 dato nulo.
* Para la columna Dirección Normalizada 8 nulos.
* Para la columna Cruce tiene 171 nulos
* Para la columna Altura tiene 567 nulos

Podemos eliminar la columna `Altura` ya que tiene una gran cantidad de nulos pero la columna altura se refiere a la misma calle donde se produce el accidente, podriamos dejar para este ejercicio unicamente la columna `calle`. El resto de columnas pueden ser manejables con esta cantidad de nulos.

In [19]:
homicidios_hechos = homicidios_hechos.drop("Altura", axis=1)

KeyError: "['Altura'] not found in axis"

In [20]:
homicidios_hechos.columns

Index(['ID', 'N_VICTIMAS', 'FECHA', 'AAAA', 'MM', 'DD', 'HORA', 'HH',
       'LUGAR_DEL_HECHO', 'TIPO_DE_CALLE', 'Calle', 'Cruce',
       'Dirección Normalizada', 'COMUNA', 'XY (CABA)', 'pos x', 'pos y',
       'PARTICIPANTES', 'VICTIMA', 'ACUSADO'],
      dtype='object')

In [21]:
#Revisamos columnas con datos duplicados

homicidios_hechos.duplicated().sum()

0

No se encuentran datos duplicados dentro del archivo

In [22]:
homicidios_hechos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 696 entries, 0 to 695
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   ID                     696 non-null    object        
 1   N_VICTIMAS             696 non-null    int64         
 2   FECHA                  696 non-null    datetime64[ns]
 3   AAAA                   696 non-null    int64         
 4   MM                     696 non-null    int64         
 5   DD                     696 non-null    int64         
 6   HORA                   696 non-null    object        
 7   HH                     696 non-null    object        
 8   LUGAR_DEL_HECHO        696 non-null    object        
 9   TIPO_DE_CALLE          696 non-null    object        
 10  Calle                  695 non-null    object        
 11  Cruce                  525 non-null    object        
 12  Dirección Normalizada  688 non-null    object        
 13  COMUN

Se observa que la primera columna con tipo de dato que necesita transformación es HH, debido a que se encuentra como objeto y deberia ser númerico.

In [23]:
#Se transforman primero a numero y despues a entero
 
homicidios_hechos["HH"] = pd.to_numeric(homicidios_hechos ["HH"], errors="coerce")
homicidios_hechos["HH"] = homicidios_hechos["HH"].astype("Int64")

In [24]:
#validamos que el cambio este ok
homicidios_hechos["HH"].unique()

<IntegerArray>
[   4,    1,    7,    0,    5,   18,   19,   15,   11,   22,   16,    9,   23,
    6,   10,   17,   12,    8,   20,   21,   14,    3,    2,   13, <NA>]
Length: 25, dtype: Int64

In [25]:
# Filtrar filas con posiciones no válidas
valid_positions = ~(homicidios_hechos['pos x'].str.contains('[^0-9.,-]') | homicidios_hechos['pos y'].str.contains('[^0-9.,-]') | (homicidios_hechos['pos x'] == '.') | (homicidios_hechos['pos y'] == '.'))

# Establecer posiciones no válidas en NaN
homicidios_hechos.loc[~valid_positions, ['pos x', 'pos y']] = np.nan

# Convertir las cadenas a tipo numérico (float)
homicidios_hechos['pos x'] = homicidios_hechos['pos x'].str.replace(',', '.').astype(float)
homicidios_hechos['pos y'] = homicidios_hechos['pos y'].str.replace(',', '.').astype(float)

In [26]:
homicidios_hechos.head(3)

ID  N_VICTIMAS      FECHA  AAAA  MM  DD      HORA  HH  \
0  2016-0001           1 2016-01-01  2016   1   1  04:00:00   4   
1  2016-0002           1 2016-01-02  2016   1   2  01:15:00   1   
2  2016-0003           1 2016-01-03  2016   1   3  07:00:00   7   

                             LUGAR_DEL_HECHO TIPO_DE_CALLE             Calle  \
0  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA  PIEDRA BUENA AV.   
1          AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ    PAZ, GRAL. AV.   
2                         AV ENTRE RIOS 2034       AVENIDA    ENTRE RIOS AV.   

                                 Cruce  \
0  FERNANDEZ DE LA CRUZ, F., GRAL. AV.   
1                  DE LOS CORRALES AV.   
2                                  NaN   

                               Dirección Normalizada  COMUNA  \
0  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...       8   
1               PAZ, GRAL. AV. y DE LOS CORRALES AV.       9   
2                                ENTRE RIOS AV. 2034       1   

                                XY (CABA)      pos x      pos y  \
0   Point (98896.78238426 93532.43437792) -58.475340 -34.687570   
1   Point (95832.05571093 95505.41641999) -58.508775 -34.669777   
2  Point (106684.29090040 99706.57687843) -58.390403 -34.631894   

    PARTICIPANTES VICTIMA    ACUSADO  
0       MOTO-AUTO    MOTO       AUTO  
1  AUTO-PASAJEROS    AUTO  PASAJEROS  
2       MOTO-AUTO    MOTO       AUTO

In [28]:
# Podemos extraer la informacion de la columna XY (CABA) ya que tiene coordenadas de geolocalización según el Diccionario de hechos y las podremos utilizar

# Función para extraer las coordenadas CABA
# Esta función toma un argumento texto, que es el texto de entrada del cual se intentarán extraer las coordenadas
# Utiliza una expresión regular (re.findall) para buscar todas las coincidencias de números decimales (que tienen un punto como separador decimal) en el texto. Estas coincidencias se almacenan en la lista coordenadas.
# return float(coordenadas[0]), float(coordenadas[1]): Si se encontraron dos coordenadas, se convierten en números de punto flotante (float) y se devuelven como una tupla de dos valores.
# Luego, se aplica esta función a la columna 'XY (CABA)' de df_siniestros utilizando el método .map(). Esto se hace para cada fila de la columna, y los resultados se almacenan en dos nuevas columnas, 'coordenada_x' y 'coordenada_y', utilizando zip(*...).

def extraer_coordenadas(texto):
    # Utilizamos una expresión regular para extraer las coordenadas
    coordenadas = re.findall(r'\d+\.\d+', texto)
    if len(coordenadas) == 2:
        return float(coordenadas[0]), float(coordenadas[1])
    else:
        return None, None

# Aplicar la función a la columna 'XY (CABA)'
homicidios_hechos['coordenada_x'], homicidios_hechos['coordenada_y'] = zip(*homicidios_hechos['XY (CABA)'].map(extraer_coordenadas))

# Filtrar los registros que no tienen coordenadas válidas
homicidios_hechos = homicidios_hechos[(homicidios_hechos['coordenada_x'].notnull()) & (homicidios_hechos['coordenada_y'].notnull())]

In [29]:
homicidios_hechos.head(3)

ID  N_VICTIMAS      FECHA  AAAA  MM  DD      HORA  HH  \
0  2016-0001           1 2016-01-01  2016   1   1  04:00:00   4   
1  2016-0002           1 2016-01-02  2016   1   2  01:15:00   1   
2  2016-0003           1 2016-01-03  2016   1   3  07:00:00   7   

                             LUGAR_DEL_HECHO TIPO_DE_CALLE  ...  \
0  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA  ...   
1          AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ  ...   
2                         AV ENTRE RIOS 2034       AVENIDA  ...   

                               Dirección Normalizada COMUNA  \
0  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...      8   
1               PAZ, GRAL. AV. y DE LOS CORRALES AV.      9   
2                                ENTRE RIOS AV. 2034      1   

                                XY (CABA)      pos x      pos y  \
0   Point (98896.78238426 93532.43437792) -58.475340 -34.687570   
1   Point (95832.05571093 95505.41641999) -58.508775 -34.669777   
2  Point (106684.29090040 99706.57687843) -58.390403 -34.631894   

    PARTICIPANTES  VICTIMA    ACUSADO   coordenada_x  coordenada_y  
0       MOTO-AUTO     MOTO       AUTO   98896.782384  93532.434378  
1  AUTO-PASAJEROS     AUTO  PASAJEROS   95832.055711  95505.416420  
2       MOTO-AUTO     MOTO       AUTO  106684.290900  99706.576878  

[3 rows x 22 columns]

In [31]:
homicidios_hechos.columns

Index(['ID', 'N_VICTIMAS', 'FECHA', 'AAAA', 'MM', 'DD', 'HORA', 'HH',
       'LUGAR_DEL_HECHO', 'TIPO_DE_CALLE', 'Calle', 'Cruce',
       'Dirección Normalizada', 'COMUNA', 'XY (CABA)', 'pos x', 'pos y',
       'PARTICIPANTES', 'VICTIMA', 'ACUSADO', 'coordenada_x', 'coordenada_y'],
      dtype='object')

In [32]:
#Podemos estandarizar los nombres de las columnas para que sea de buena visibilidad a la hora de trabajar el BI

#Cambio la primer letra a mayúscula
homicidios_hechos.columns = [x.capitalize() for x in homicidios_hechos.columns]
# Reemplazo guiones por espacios
homicidios_hechos.columns = homicidios_hechos.columns.str.replace("_", " ")
#Renombra columnas 
homicidios_hechos= homicidios_hechos.rename(columns={"N victimas": "Num víctimas",
                                            "Aaaa": "Año",
                                            "Mm":"Mes",
                                            "Dd": "Día",
                                            "Hora": "Hora completa",
                                            "Hh": "Hora"}) 

homicidios_hechos.columns

Index(['Id', 'Num víctimas', 'Fecha', 'Año', 'Mes', 'Día', 'Hora completa',
       'Hora', 'Lugar del hecho', 'Tipo de calle', 'Calle', 'Cruce',
       'Dirección normalizada', 'Comuna', 'Xy (caba)', 'Pos x', 'Pos y',
       'Participantes', 'Victima', 'Acusado', 'Coordenada x', 'Coordenada y'],
      dtype='object')

In [33]:
#Podemos agregar una nueva columna con los dias de la semana en Nombre remplazando la que esta Numerica con el fin de tener una buena visualizacion en al BI

#Se crea una columna nueva `Día semana` que contenga el nombre del día de la semana
homicidios_hechos['Día semana'] = homicidios_hechos['Fecha'].dt.strftime('%A')

#Se crea un diccionario para modificar a español los nombres de los dás de la semana y lo aplico a la columna nueva
dic={  'Friday':'Viernes',
       'Saturday': 'Sabado',
       'Sunday': 'Domingo',  
       'Thursday': 'Jueves',
        'Monday': 'Lunes',
        'Wednesday': 'Miercoles',
       'Tuesday': 'Martes'}
homicidios_hechos['Día semana'] =homicidios_hechos['Día semana'].replace(dic)

In [34]:
homicidios_hechos.head(5)

Id  Num víctimas      Fecha   Año  Mes  Día Hora completa  Hora  \
0  2016-0001             1 2016-01-01  2016    1    1      04:00:00     4   
1  2016-0002             1 2016-01-02  2016    1    2      01:15:00     1   
2  2016-0003             1 2016-01-03  2016    1    3      07:00:00     7   
3  2016-0004             1 2016-01-10  2016    1   10      00:00:00     0   
4  2016-0005             1 2016-01-21  2016    1   21      05:20:00     5   

                             Lugar del hecho Tipo de calle  ... Comuna  \
0  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA  ...      8   
1          AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ  ...      9   
2                         AV ENTRE RIOS 2034       AVENIDA  ...      1   
3      AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA  ...      8   
4  AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA  ...      1   

                                 Xy (caba)      Pos x      Pos y  \
0    Point (98896.78238426 93532.43437792) -58.475340 -34.687570   
1    Point (95832.05571093 95505.41641999) -58.508775 -34.669777   
2   Point (106684.29090040 99706.57687843) -58.390403 -34.631894   
3    Point (99840.65224780 94269.16534422) -58.465039 -34.680930   
4  Point (106980.32827929 100752.16915795) -58.387183 -34.622466   

    Participantes  Victima    Acusado   Coordenada x   Coordenada y Día semana  
0       MOTO-AUTO     MOTO       AUTO   98896.782384   93532.434378    Viernes  
1  AUTO-PASAJEROS     AUTO  PASAJEROS   95832.055711   95505.416420     Sabado  
2       MOTO-AUTO     MOTO       AUTO  106684.290900   99706.576878    Domingo  
3         MOTO-SD     MOTO         SD   99840.652248   94269.165344    Domingo  
4  MOTO-PASAJEROS     MOTO  PASAJEROS  106980.328279  100752.169158     Jueves  

[5 rows x 23 columns]

## Transformacion de la data `homicidios_victimas`

In [35]:
homicidios_victimas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 717 entries, 0 to 716
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   ID_hecho             717 non-null    object        
 1   FECHA                717 non-null    datetime64[ns]
 2   AAAA                 717 non-null    int64         
 3   MM                   717 non-null    int64         
 4   DD                   717 non-null    int64         
 5   ROL                  717 non-null    object        
 6   VICTIMA              717 non-null    object        
 7   SEXO                 717 non-null    object        
 8   EDAD                 717 non-null    object        
 9   FECHA_FALLECIMIENTO  717 non-null    object        
dtypes: datetime64[ns](1), int64(3), object(6)
memory usage: 56.1+ KB


In [36]:
#Podemos cambiar los tipos de datos de las columnas FECHA_FALLECIMIENTO y EDAD con el fin de manejar mejor los datos

#Cambio la columna a tipo datetime
homicidios_victimas["FECHA_FALLECIMIENTO"] = pd.to_datetime(homicidios_victimas["FECHA_FALLECIMIENTO"], format='%Y-%m-%d', errors='coerce')

In [37]:
#Una vez que se paso el tipo de dato a datetime que es una fecha y hora completa se utiliza este codigo para que solo se quede en formato fecha sin hora
homicidios_victimas["FECHA_FALLECIMIENTO"] = homicidios_victimas["FECHA_FALLECIMIENTO"].dt.date

#Luego se reemplazan las fechas en cero con espacios vacios
homicidios_victimas["FECHA_FALLECIMIENTO"] = homicidios_victimas ["FECHA_FALLECIMIENTO"].replace(pd.NaT,"")

In [42]:
#Cambio la columna edad a numerico y luego a tipo entero
homicidios_victimas ["EDAD"] = pd.to_numeric(homicidios_victimas["EDAD"], errors="coerce")
homicidios_victimas ["EDAD"] = homicidios_victimas["EDAD"].astype("Int64")

In [43]:
#Utilizo Numpy para reemplazar por NaN los sin datos o vacios y luego paso a float para que cuando quiera aplicar estadisticas no tenga errores debido a esos datos
homicidios_victimas["EDAD"] = homicidios_victimas["EDAD"].replace([" ","SD"], np.nan)

In [44]:
homicidios_victimas["EDAD"] = homicidios_victimas["EDAD"].fillna(0.0)

In [45]:
#Validamos el cambio
homicidios_victimas ["EDAD"].unique()

<IntegerArray>
[19, 70, 30, 18, 29, 22, 16, 59, 65, 34, 41, 50, 38, 21, 52, 36, 20, 54,  0,
 56, 24, 78, 79, 26, 57, 37, 58, 23, 60, 42, 53, 51, 40, 87, 76, 75, 35, 80,
 43, 45, 67, 27, 55, 49, 81, 25, 33, 46, 83, 39, 28,  7, 48,  4, 82, 32, 17,
 47, 61, 10, 95, 73, 84, 66, 85,  1, 15, 13, 77, 44, 31, 62, 74, 71, 11, 86,
 69, 72, 12, 63, 92, 68, 91, 64,  5, 88]
Length: 86, dtype: Int64

In [46]:
homicidios_victimas.head(3)

ID_hecho      FECHA  AAAA  MM  DD        ROL VICTIMA       SEXO  EDAD  \
0  2016-0001 2016-01-01  2016   1   1  CONDUCTOR    MOTO  MASCULINO    19   
1  2016-0002 2016-01-02  2016   1   2  CONDUCTOR    AUTO  MASCULINO    70   
2  2016-0003 2016-01-03  2016   1   3  CONDUCTOR    MOTO  MASCULINO    30   

  FECHA_FALLECIMIENTO  
0          2016-01-01  
1          2016-01-02  
2          2016-01-03

Nulos y Duplicados

In [47]:
homicidios_victimas.isnull().sum()

ID_hecho               0
FECHA                  0
AAAA                   0
MM                     0
DD                     0
ROL                    0
VICTIMA                0
SEXO                   0
EDAD                   0
FECHA_FALLECIMIENTO    0
dtype: int64

Vemos que no hay ningun dato nulo

In [48]:
#Estandarizamos los nombres de las Columnas
# Se coloca la primera en mayúscula
homicidios_victimas.columns = [x.capitalize() for x in homicidios_victimas.columns]
# Se reemplazan los guiones por espacios
homicidios_victimas.columns = homicidios_victimas.columns.str.replace('_', ' ')
# Se renombran algunas columnas
homicidios_victimas = homicidios_victimas.rename(columns={'Id hecho': 'Id',
                                                      'Aaaa':'Año',
                                                      'Mm':'Mes',
                                                      'Dd':'Día',
                                                      'Victima':'Víctima'})
homicidios_victimas.columns

Index(['Id', 'Fecha', 'Año', 'Mes', 'Día', 'Rol', 'Víctima', 'Sexo', 'Edad',
       'Fecha fallecimiento'],
      dtype='object')

Por ultimo podemos crear una columna de rango etario para mejor analisis en el BI

In [50]:
# Se utiliza la función `pd.cut()`, que permite agrupar valores numéricos en intervalos o categorías


# Definir los límites de los rangos etarios y las etiquetas
bins = [0, 20, 40, 60, 80, 100]
labels = ['Menos de 20', ' Entre 21- 40', 'Entre 41-60', 'Entre 61-80', 'Más de 80']

# Utilizar pd.cut() para crear la columna de rango etario
homicidios_victimas['Rango_etario'] = pd.cut(homicidios_victimas['Edad'], bins=bins, labels=labels, right=False)

homicidios_victimas.head(3)

Id      Fecha   Año  Mes  Día        Rol Víctima       Sexo  Edad  \
0  2016-0001 2016-01-01  2016    1    1  CONDUCTOR    MOTO  MASCULINO    19   
1  2016-0002 2016-01-02  2016    1    2  CONDUCTOR    AUTO  MASCULINO    70   
2  2016-0003 2016-01-03  2016    1    3  CONDUCTOR    MOTO  MASCULINO    30   

  Fecha fallecimiento   Rango_etario  
0          2016-01-01    Menos de 20  
1          2016-01-02    Entre 61-80  
2          2016-01-03   Entre 21- 40

Con el fin de unificar ambos DataSets en uno solo podemos eliminar columnas repetidas, para este caso podemos eliminar del dataset homicidios_vistimas la columna fecha, Año, Mes y Dia ya que traen los mismos datos del dataSet Homicidios_hechos

In [51]:
#Se busca un registro para comparar con el mismo registro en el otro dataset y vemos que es la misma persona
homicidios_hechos[homicidios_hechos["Id"]=="2019-0012"]

Id  Num víctimas      Fecha   Año  Mes  Día Hora completa  Hora  \
429  2019-0012             1 2019-02-17  2019    2   17      03:16:00     3   

                        Lugar del hecho Tipo de calle  ... Comuna  \
429  CASTAÑARES AV. Y GORDILLO, TIMOTEO       AVENIDA  ...      8   

                                 Xy (caba)      Pos x      Pos y  \
429  Point (98072.98959434 94742.80804806) -58.484327 -34.676658   

    Participantes  Victima  Acusado  Coordenada x  Coordenada y Día semana  
429     MOTO-AUTO     MOTO     AUTO  98072.989594  94742.808048    Domingo  

[1 rows x 23 columns]

In [52]:
homicidios_victimas[homicidios_victimas["Id"]=="2019-0012"]

Id      Fecha   Año  Mes  Día        Rol Víctima       Sexo  Edad  \
447  2019-0012 2019-02-17  2019    2   17  CONDUCTOR    MOTO  MASCULINO    31   

    Fecha fallecimiento   Rango_etario  
447          2019-02-20   Entre 21- 40

In [53]:
# Se eliminan las columnas repetidas
homicidios_victimas = homicidios_victimas.drop(['Fecha', 'Año', 'Mes', 'Día', 'Víctima'], axis=1)
homicidios_victimas.columns

Index(['Id', 'Rol', 'Sexo', 'Edad', 'Fecha fallecimiento', 'Rango_etario'], dtype='object')

## Unión de los DataSets